src/shogun/base/base_types.h

```diff
@@ -68,6 +70,121 @@ namespace shogun
 	                  const char*, typename std::decay<T>::type>::value>
 	{
 	};
+
+	// Primitive types
+	enum class TYPE
+	{
+		T_BOOL = 1,
+		T_CHAR = 2,
+		T_INT8 = 3,
+		T_UINT8 = 4,
+		T_INT16 = 5,
+		T_UINT16 = 6,
+		T_INT32 = 7,
+		T_UINT32 = 8,
+		T_INT64 = 9,
+		T_UINT64 = 10,
+		T_FLOAT32 = 11,
+		T_FLOAT64 = 12,
+		T_FLOATMAX = 13,
+		T_SGOBJECT = 14,
+		T_COMPLEX128 = 15,
+		T_SGVECTOR_FLOAT32 = 16,
+		T_SGVECTOR_FLOAT64 = 17,
+		T_SGVECTOR_FLOATMAX = 18,
+		T_SGVECTOR_INT32 = 19,
+		T_SGVECTOR_INT64 = 20,
+		T_SGMATRIX_FLOAT32 = 21,
+		T_SGMATRIX_FLOAT64 = 22,
+		T_SGMATRIX_FLOATMAX = 23,
+		T_SGMATRIX_INT32 = 24,
+		T_SGMATRIX_INT64 = 25,
+		T_UNDEFINED = 26
+	};
```

Comment:
Do you think it would be worth adding all the combinations of containers and primitive types, or some macro that would do that? Because right now stuff like `SGVector<bool>` isn't included (whether or not we actually use it is another question...)

Reply:
I guess it would make sense if we were to use it somewhere in Shogun. For instance, if we find a machine which needs to emit `SGVector<bool>` then we will add it.